File: java/src/patterns/creational/singleton/BillPughSingleton.java
package patterns.creational.singleton;

class BillPughSingleton {
    // private constructor prevents object instantiation
    private BillPughSingleton() {
        // init logic here
    }

    // static class to hold the single instance
    private static class SingletonHelper {
        private static final BillPughSingleton INSTANCE = new BillPughSingleton();
    }

    public static BillPughSingleton getInstance() {
        return SingletonHelper.INSTANCE;
    }
}

File: python/src/patterns/creational/singleton/bill_pugh_singleton.py
class BillPughSingleton:
    class _SingletonHelper:
        INSTANCE = None

    def __init__(self) -> None:
        # prevent multiple instantiation by raising an exception
        if BillPughSingleton._SingletonHelper.INSTANCE is not None:
            raise Exception("Cannot instantiate directly.")
        # init logic here

    @staticmethod
    def get_instance() -> "BillPughSingleton":
        return BillPughSingleton._SingletonHelper.INSTANCE
